app/src/ui/history/compare.tsx

```diff
@@ -78,32 +78,62 @@ export class CompareSidebar extends React.Component<
     this.state = {
       focusedBranch: null,
       filterText,
-      showBranchList: props.initialShowBranchList,
+      showBranchList: props.shouldShowBranchesList,
       selectedCommit: null,
     }
   }
 
   public componentWillReceiveProps(nextProps: ICompareSidebarProps) {
-    const hasFormStateChanged =
-      nextProps.compareState.formState.kind !==
-      this.props.compareState.formState.kind
-
     const newFormState = nextProps.compareState.formState
-
-    if (hasFormStateChanged && newFormState.kind === ComparisonView.None) {
-      // the comparison form should be reset to its default state
-      this.setState({ filterText: '', focusedBranch: null })
+    const oldFormState = this.props.compareState.formState
+
+    if (
+      newFormState.kind !== oldFormState.kind &&
+      newFormState.kind === ComparisonView.None
+    ) {
+      // reset form to it's default state
+      this.setState(
+        {
+          filterText: '',
+          focusedBranch: null,
+          showBranchList: nextProps.shouldShowBranchesList,
+        },
+        () => {
+          // ensure filter text behaviour matches the prop value
+          if (this.textbox) {
+            if (nextProps.shouldShowBranchesList) {
+              this.textbox.focus()
+            } else {
+              this.textbox.blur()
+            }
+          }
+        }
+      )
       return
     }
 
-    if (newFormState.kind !== ComparisonView.None) {
-      // ensure the filter text is in sync with the comparison branch
-      const branch = newFormState.comparisonBranch
+    if (
+      newFormState.kind !== ComparisonView.None &&
+      oldFormState.kind !== ComparisonView.None
+    ) {
+      const oldBranch = oldFormState.comparisonBranch
+      const newBranch = newFormState.comparisonBranch
+
+      if (oldBranch.name !== newBranch.name) {
+        // ensure the filter text is in sync with the comparison branch
+        this.setState({
+          filterText: newBranch.name,
+          focusedBranch: newBranch,
+        })
+      }
+    }
 
-      this.setState({
-        filterText: branch.name,
-        focusedBranch: branch,
-      })
+    if (
```

Comment:
the only way this happens is if `this.props.shouldShowBranchesList` is `false` and `nextProps.shouldShowBranchesList` is `true` right? maybe this can get folded in with what's on line 139? seems like we're just `setState`-ing to be `nextProps.shouldShowBranchesList` if it's different from `this.props.shouldShowBranchesList`

is there a case I'm missing?

Reply:
This is kind of related to the issue reported https://github.com/desktop/desktop/pull/4743#issuecomment-390822487, and it's related to managing both prop changes and state. This check is helpful for the `Compare to Branch` menu item, but the equivalent checks for `Show History` (`true` -> `false` isn't there).

> maybe this can get folded in with what's on line 139?

I'd rather lift up this state and then figure out what to do with the `sidebarHasFocusWithin` check as part of that work (I think it can go away too).